File: 2023/kotlin/src/Day21.kt
private data class Garden(val totalSize: Int, val plots: Set<Point>, val start: Point) {
    fun totalPlotsLandedOnAfterSteps(steps: Int): Long {
        var landedOn = setOf(start)

        repeat(steps) {
            landedOn = buildSet {
                landedOn.forEach { plot ->
                    plot.cardinalNeighbors().forEach { nextPlot ->
                        if (Point(nextPlot.x.mod(totalSize), nextPlot.y.mod(totalSize)) in plots) add(nextPlot)
                    }
                }
            }
        }

        return landedOn.size.toLong()
    }

    companion object {
        fun of(input: String): Garden {
            val size = input.lines().size
            var start = Point.ORIGIN
            val plots = buildSet {
                input.lines().forEachIndexed { y, row ->
                    row.forEachIndexed { x, value ->
                        if (value == 'S') start = Point(x, y)
                        if (value != '#') add(Point(x, y))
                    }
                }
            }

            return Garden(size, plots, start)
        }
    }
}

private fun part1(input: String) =
    Garden.of(input).totalPlotsLandedOnAfterSteps(64)

private fun part2(input: String): Long {
    val garden = Garden.of(input)

    val target = (26_501_365 - 65) / 131;
    val ys = listOf(
        garden.totalPlotsLandedOnAfterSteps(65),
        garden.totalPlotsLandedOnAfterSteps(65 + 131),
        garden.totalPlotsLandedOnAfterSteps(65 + 2 * 131),
    )

    // Simplified Lagrange's interpolation formula
    // https://www.wolframalpha.com/input?i=quadratic+fit+calculator&assumption=%7B%22F%22%2C+%22QuadraticFitCalculator%22%2C+%22data3x%22%7D+-%3E%22%7B0%2C+1%2C+2%7D%22&assumption=%7B%22F%22%2C+%22QuadraticFitCalculator%22%2C+%22data3y%22%7D+-%3E%22%7B3835%2C+34125%2C+94603%7D%22
    return ys
        .mapIndexed { idx, y ->
            ys.indices.filterNot { it == idx }.fold(y) { acc, x ->
                acc * (target - x) / (idx - x)
            }
        }
        .sum()
}

fun main() {
    val input = readInput("Day21")
    part1(input).println()
    part2(input).println()
}

File: 2023/swift/AdventOfCode2023/Sources/Day21.swift
infix operator %% : MultiplicationPrecedence

func %% <T: BinaryInteger>(lhs: T, rhs: T) -> T {
  return (lhs % rhs + rhs) % rhs
}

struct Day21: AdventDay {
  var input: String

  enum StepDirection {
    case up, down, left, right
  }

  struct StepPoint: Hashable {
    let y: Int, x: Int

    init(_ y: Int, _ x: Int) {
      self.y = y
      self.x = x
    }

    func step(_ direction: StepDirection) -> Self {
      switch direction {
      case .up: return StepPoint(y - 1, x)
      case .down: return StepPoint(y + 1, x)
      case .left: return StepPoint(y, x - 1)
      case .right: return StepPoint(y, x + 1)
      }
    }

    func neighbours() -> [StepPoint] {
      [StepDirection.up, .down, .left, .right].map { self.step($0) }
    }
  }

  struct Garden {
    let size: Int
    let plots: Set<StepPoint>
    let start: StepPoint

    func totalPlotsLandedOnAfter(steps: Int) -> Int {
      var landedOn = Set([start])

      for _ in 1...steps {
        landedOn = Set(
          landedOn.flatMap { plot in
            plot.neighbours().compactMap { nextPlot in
              plots.contains(StepPoint(nextPlot.y %% size, nextPlot.x %% size)) ? nextPlot : nil
            }
          })
      }

      return landedOn.count
    }

    static func from(_ input: String) -> Self {
      var start = StepPoint(0, 0)
      let plots = Set(
        input.lines().enumerated().flatMap { y, row in
          row.enumerated().compactMap { x, el in
            if el == "S" { start = StepPoint(y, x) }
            return el != "#" ? StepPoint(y, x) : nil
          }
        })

      return Garden(size: input.lines().count, plots: plots, start: start)
    }
  }

  func part1() -> Any {
    Garden.from(input).totalPlotsLandedOnAfter(steps: 64)
  }

  func part2() -> Any {
    let garden = Garden.from(input)

    let target = (26_501_365 - 65) / 131
    let ys = [
      garden.totalPlotsLandedOnAfter(steps: 65),
      garden.totalPlotsLandedOnAfter(steps: 65 + 131),
      garden.totalPlotsLandedOnAfter(steps: 65 + 2 * 131),
    ]

    // Simplified Lagrange's interpolation formula
    // https://www.wolframalpha.com/input?i=quadratic+fit+calculator&assumption=%7B%22F%22%2C+%22QuadraticFitCalculator%22%2C+%22data3x%22%7D+-%3E%22%7B0%2C+1%2C+2%7D%22&assumption=%7B%22F%22%2C+%22QuadraticFitCalculator%22%2C+%22data3y%22%7D+-%3E%22%7B3835%2C+34125%2C+94603%7D%22
    return
      ys
      .enumerated()
      .map { idx, y in
        (0..<ys.count)
          .filter { $0 != idx }
          .reduce(y) { acc, x in acc * (target - x) / (idx - x) }
      }
      .sum()
  }
}
